### PR TITLE
New metaconfig common classes

### PR DESCRIPTION
**Pull request by bbaldino**
This PR includes some common code oriented around using the new metaconfig lib.  It defines classes which:
1) Implement a metaconfig `ConfigSource` bridge to the typesafe config library (`NewTypesafeConfigSource`)
1) Implement a metaconfig `ConfigSource` bridge to the legacy `ConfigurationService` class (`ConfigurationServiceConfigSource`)
1) House the instances of these configurations (`NewJitsiConfig`)

~Some classes are prefixed with `New` to avoid clashing with the existing ones--I held off on ripping those out in this PR because I think we'll port the config properties over in JVB one-by-one, so can rip them out once they're no longer used (at which point the names of these classes will have the `New` prefix removed).~ This PR has been updated to rip out the old classes.

### CODE REVIEW

File: jicoco-kotlin/src/main/kotlin/org/jitsi/config/AbstractReadOnlyConfigurationService.kt [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+    override fun logConfigurationProperties(passwordPattern: String) {
+        val regex = Regex(passwordPattern).takeIf { passwordPattern.isNotEmpty() }
+
+        for (name in allPropertyNames) {
+            var value = getProperty(name) ?: continue
+
+            if (regex?.matches(name) == true) {
+                value = "**********"
+            }
+            // TODO(brian): which logger to use here?
```
Comment (JonathanLennox) [inline jicoco-kotlin/src/main/kotlin/org/jitsi/config/AbstractReadOnlyConfigurationService.kt:84]: Resolve this

File: jicoco-kotlin/src/main/kotlin/org/jitsi/config/TypesafeConfigSource.kt [hunk 2]
```diff
@@ -17,96 +17,62 @@
...
 import kotlin.reflect.KClass
+import kotlin.reflect.KType
+import kotlin.reflect.full.isSubtypeOf
+import kotlin.reflect.typeOf
 
 /**
- * A [ConfigSource] which reads from a typesafe [Config] object.
+ * A [ConfigSource] implementation backed by a [Config] instance.
+ *
+ * (Should be renamed to TypesafeConfigSource once the current one has been removed)
```
Comment (JonathanLennox) [inline jicoco-kotlin/src/main/kotlin/org/jitsi/config/TypesafeConfigSource.kt:32]: Comment can be removed now

File: jicoco-kotlin/src/main/kotlin/org/jitsi/config/JitsiConfig.kt [hunk 3]
```diff
@@ -17,38 +17,42 @@
...
     companion object {
-        private val logger = LoggerImpl(JitsiConfig::class.qualifiedName)
-        val newConfig: ConfigSource = JitsiConfigFactory.newConfigSupplier()
-        val legacyConfig: ConfigSource = JitsiConfigFactory.legacyConfigSupplier()
-        @JvmStatic
-        val legacyConfigShim = JitsiConfigFactory.legacyConfigurationServiceShimSupplier()
+        /**
+         * A [ConfigSource] loaded via [ConfigFactory].
+         */
+        val TypesafeConfig: ConfigSource = TypesafeConfigSource("typesafe config", ConfigFactory.load())
```
Comment (bgrozev) [inline jicoco-kotlin/src/main/kotlin/org/jitsi/config/JitsiConfig.kt:34]: Is it a convention that the names of `val`s in the companion object start with a capital letter?

File: jicoco-kotlin/src/main/kotlin/org/jitsi/config/ReadOnlyConfigurationService.kt [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+    init {
+        reloadConfiguration()
+    }
+
+    override fun reloadConfiguration() {
+        val scHomeDirLocation = System.getenv(ConfigurationService.PNAME_SC_HOME_DIR_LOCATION)
+            ?: System.getProperty(ConfigurationService.PNAME_SC_HOME_DIR_LOCATION)
+            ?: run {
+                println("${ConfigurationService.PNAME_SC_HOME_DIR_LOCATION} not set")
```
Comment (bgrozev) [inline jicoco-kotlin/src/main/kotlin/org/jitsi/config/ReadOnlyConfigurationService.kt:40]: Can we use a logger? The parent class has one

File: jicoco-kotlin/src/main/kotlin/org/jitsi/config/ReadOnlyConfigurationService.kt [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+                return
+            }
+        val scHomeDirName = System.getenv(ConfigurationService.PNAME_SC_HOME_DIR_NAME)
+            ?: System.getProperty(ConfigurationService.PNAME_SC_HOME_DIR_NAME)
+            ?: run {
+                println("${ConfigurationService.PNAME_SC_HOME_DIR_NAME} not set")
+                return
+            }
+        val fileName = "sip-communicator.properties"
+        println("loading config file ${Paths.get(scHomeDirLocation, scHomeDirName, fileName)}")
```
Comment (bgrozev) [inline jicoco-kotlin/src/main/kotlin/org/jitsi/config/ReadOnlyConfigurationService.kt:50]: nit: Save the result of `Paths.get`

File: jicoco-kotlin/pom.xml [hunk 6]
```diff
@@ -152,5 +165,9 @@
        </snapshots>
        <url>https://github.com/jitsi/jitsi-maven-repository/raw/master/snapshots/</url>
     </repository>
+   <repository>
+        <id>jitpack.io</id>
```
[the remark this answers is not in the document]
Comment (bbaldino) [reply jicoco-kotlin/pom.xml:169]: `jitsi-metaconfig` is currently hosted there.

File: jicoco-kotlin/src/test/kotlin/org/jitsi/config/TypesafeConfigSourceTest.kt [hunk 7]
```diff
@@ -16,106 +16,113 @@
...
+                    """
+                        obj = {
+                            num = 42
+                            str = "hello"
+                        }
+                    """.trimIndent()
                 }
-                getValue<Duration>("some.duration.value") shouldBe Duration.ofMillis(500)
+                val obj = getValue<ConfigObject>("obj")
+                obj["num"]!!.unwrapped() as Number shouldBe 42
```
[the remark this answers is not in the document]
Comment (bbaldino) [reply jicoco-kotlin/src/test/kotlin/org/jitsi/config/TypesafeConfigSourceTest.kt:80]: I thought that would be less accurate, thinking that a `ConfigObject` could unwrap into different things, but as part of looking into that I realized I was wrong :) So no.  I'll change it.

File: jicoco-kotlin/src/test/kotlin/org/jitsi/config/AbstractReadOnlyConfigurationServiceTest.kt [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+    override fun isolationMode(): IsolationMode? = IsolationMode.InstancePerLeaf
+
+    private val config = TestReadOnlyConfigurationService()
+
+    init {
+        "retrieving a property" {
+            "present in both the config and the system" {
+                config["some.prop"] = "42"
+                withSystemProperties("some.prop" to "43") {
+                    should("use the system property") {
```
[the remark this answers is not in the document]
Comment (bbaldino) [reply jicoco-kotlin/src/test/kotlin/org/jitsi/config/AbstractReadOnlyConfigurationServiceTest.kt:35]: Nice catch. Fixed the description.

File: jicoco-kotlin/src/main/kotlin/org/jitsi/config/TypesafeConfigSource.kt [hunk 9]
```diff
@@ -17,96 +17,60 @@
...
+     * Translate [com.typesafe.config.ConfigException]s into [ConfigException]
+     */
+    private fun wrap(block: (String) -> Any): (String) -> Any {
+        return { key ->
+            try {
+                block(key)
+            } catch (e: com.typesafe.config.ConfigException.Missing) {
+                throw ConfigException.UnableToRetrieve.NotFound("Key '$key' not found in source '$name'")
+            } catch (e: com.typesafe.config.ConfigException.WrongType) {
+                throw ConfigException.UnableToRetrieve.WrongType("Key '$key' in source '$name': ${e.message}")
```
[the remark this answers is not in the document]
Comment (bbaldino) [reply jicoco-kotlin/src/main/kotlin/org/jitsi/config/TypesafeConfigSource.kt:71]: `WrongType` is handled here.  I was thinking I didn't want to handle more than what I'd seen...but it's probably a good idea to have a catch-all.  I'll add one.

File: jicoco-kotlin/src/main/kotlin/org/jitsi/config/AbstractReadOnlyConfigurationService.kt [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+        getString(propertyName)?.toLong() ?: defaultValue
+
+    override fun getAllPropertyNames(): MutableList<String> =
+        properties.keys.map { it as String}.toMutableList()
+
+    override fun getProperty(propertyName: String): Any? {
+        return properties[propertyName] ?: System.getProperty(propertyName)
+    }
+
+    override fun getPropertyNamesByPrefix(prefix: String, exactPrefixMatch: Boolean): MutableList<String> {
```
[the remark this answers is not in the document]
Comment (bbaldino) [reply jicoco-kotlin/src/main/kotlin/org/jitsi/config/AbstractReadOnlyConfigurationService.kt:63]: Added one

File: jicoco-kotlin/src/main/kotlin/org/jitsi/config/ReadOnlyConfigurationService.kt (hunk 4 above)
Comment (bbaldino) [reply jicoco-kotlin/src/main/kotlin/org/jitsi/config/ReadOnlyConfigurationService.kt:40]: Oops, nice catch. Fixed.

File: jicoco-kotlin/src/main/kotlin/org/jitsi/config/ConfigurationServiceConfigSource.kt [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+import kotlin.reflect.typeOf
+
+/**
+ * A [ConfigSource] implementation which is backed by a [ConfigurationService] instance.
+ */
+class ConfigurationServiceConfigSource(
+    override val name: String,
+    private val config: ConfigurationService
+) : ConfigSource {
+
```
[the remark this answers is not in the document]
Comment (JonathanLennox) [reply jicoco-kotlin/src/main/kotlin/org/jitsi/config/ConfigurationServiceConfigSource.kt:32]: (Answer: no, we're not.  See https://github.com/jitsi/jicoco/pull/90, which I'll rebase after this goes in.)

File: jicoco-kotlin/src/main/kotlin/org/jitsi/config/JitsiConfig.kt (hunk 3 above)
Comment (bbaldino) [reply jicoco-kotlin/src/main/kotlin/org/jitsi/config/JitsiConfig.kt:34]: I took this style from the kotlin guys from how they defined [Dispatchers](https://github.com/Kotlin/kotlinx.coroutines/blob/master/kotlinx-coroutines-core/jvm/src/Dispatchers.kt#L33)...I think it's more differentiating the default value.  I don't feel strongly though.

**Comment by bbaldino** [comment]
jenkins, test this please

**Pull request merged**